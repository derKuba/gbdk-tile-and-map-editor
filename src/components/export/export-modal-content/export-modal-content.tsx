import { Component, h, State, Prop } from '@stencil/core';

import Tunnel from '../../../provider/StateProvider';

enum CONTENT_TYPES {
    SELECTOR = 0,
    CODE_OUTPUT = 1,
}

@Component({
    tag: 'mvjp-export-modal-content',
    styleUrl: 'export-modal-content.css',
    shadow: true,
})
export class ExportModalContent {
    /**
     *  id of the display modal content
     */
    @State() activeIndex: number = CONTENT_TYPES.SELECTOR;

    /**
     *  state handler for choosen checkbox
     */
    @State() selectedFields: number[] = [];

    /**
     *  the painted matrix
     */
    @Prop() matrix: number[][][];

    private setIndex = index => {
        this.activeIndex = index;
    };

    private handleTileClick = position => {
        if (this.selectedFields[position] === position) {
            this.selectedFields[position] = undefined;
        } else {
            this.selectedFields[position] = position;
        }

        this.selectedFields = [...this.selectedFields];
    };

    render() {
        const selectedFields: any = this.selectedFields
            .filter(index => index !== undefined)
            .map(index => this.matrix[index]);

        if (this.activeIndex === CONTENT_TYPES.SELECTOR) {
            return (
                <mvjp-export-field-selector
                    navigateTo={() => this.setIndex(CONTENT_TYPES.CODE_OUTPUT)}
                    handleTileClick={this.handleTileClick}
                    selectedFields={this.selectedFields}
                    matrix={this.matrix}
                />
            );
        } else if (this.activeIndex === CONTENT_TYPES.CODE_OUTPUT) {
            return (
                <mvjp-export-code-output
                    navigateTo={() => this.setIndex(CONTENT_TYPES.SELECTOR)}
                    fields={selectedFields}
                />
            );
        }
    }
}
Tunnel.injectProps(ExportModalContent, ['matrix']); // eslint-disable-line @stencil/ban-side-effects
