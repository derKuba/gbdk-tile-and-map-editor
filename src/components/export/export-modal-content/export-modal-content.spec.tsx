import { newSpecPage } from '@stencil/core/dist/testing';
import { h } from '@stencil/core';

import { ExportModalContent } from './export-modal-content';

describe('mvjp-export-modal-content', () => {
    it('builds', () => {
        expect(new ExportModalContent()).toBeTruthy();
    });

    it('matches snapshot', async () => {
        const page = await newSpecPage({
            components: [ExportModalContent],
            template: () => {
                return <mvjp-export-modal-content />;
            },
        });

        expect(page.root).toMatchSnapshot();

        page.rootInstance.setIndex(1);

        await page.waitForChanges();

        expect(page.root).toMatchSnapshot();
    });

    it('should handle tile click', async () => {
        const modalContent = new ExportModalContent();
        modalContent.selectedFields = [0, 1, undefined];
        modalContent.matrix = [
            [
                [
                    [0, 1, 2],
                    [0, 1, 2],
                    [0, 1, 2],
                ],
                [
                    [0, 1, 2],
                    [0, 1, 2],
                    [0, 1, 2],
                ],
                [
                    [0, 1, 2],
                    [0, 1, 2],
                    [0, 1, 2],
                ],
            ],
        ];

        modalContent.handleTileClick(2);

        expect(modalContent.selectedFields).toEqual([0, 1, 2]);

        modalContent.handleTileClick(1);
        expect(modalContent.selectedFields).toEqual([0, undefined, 2]);
    });
});
