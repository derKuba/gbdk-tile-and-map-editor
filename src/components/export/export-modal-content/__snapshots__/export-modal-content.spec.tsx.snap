// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mvjp-export-modal-content matches snapshot 1`] = `
<mvjp-export-modal-content>
  <mock:shadow-root>
    <mvjp-export-field-selector></mvjp-export-field-selector>
  </mock:shadow-root>
</mvjp-export-modal-content>
`;

exports[`mvjp-export-modal-content matches snapshot 2`] = `
<mvjp-export-modal-content>
  <mock:shadow-root>
    <mvjp-export-code-output></mvjp-export-code-output>
  </mock:shadow-root>
</mvjp-export-modal-content>
`;
